packages/core/src/new-api/types/journal.ts

```diff
@@ -0,0 +1,197 @@
+import { ArgumentType, FutureType } from "./module";
+
+/**
+ * Store a deployments execution state as a transaction log.
+ *
+ * @beta
+ */
+export interface Journal {
+  record(message: JournalableMessage): Promise<void>;
```

Comment:
I don't think this should be async. To ensure some reasonable durability of the journal we should append synchronously making sure that everything is flushed before continuing the execution. I've already did some research about how to achieve this in node.

Reply:
I have added this as a todo in the restartability issue: https://github.com/NomicFoundation/ignition/pull/255#discussion_r1228729914